File: shared/src/jvmTest/kotlin/space/denhan/backup/BackupTest.kt
package space.denhan.backup

import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.first
import kotlinx.coroutines.test.runTest
import kotlinx.datetime.LocalDate
import space.denhan.data.DriverFactory
import space.denhan.data.ItemRepository
import space.denhan.data.createDatabase
import space.denhan.domain.Carrier
import space.denhan.domain.Cycle
import space.denhan.domain.DateSource
import space.denhan.domain.GroupKind
import space.denhan.domain.HandledEvent
import space.denhan.domain.ItemGroup
import space.denhan.domain.Kind
import space.denhan.domain.SimProfile
import space.denhan.domain.TrackedItem
import space.denhan.domain.TriState
import kotlin.test.Test
import kotlin.test.assertContains
import kotlin.test.assertEquals
import kotlin.test.assertTrue

class BackupTest {

    private fun d(iso: String) = LocalDate.parse(iso)
    private fun repo() = ItemRepository(createDatabase(DriverFactory()), Dispatchers.Default)

    private val group = ItemGroup("sim1", "SIM Viettel 0912 345 678", GroupKind.SIM)

    private val sim = SimProfile(
        groupId = "sim1",
        carrier = Carrier.VIETTEL,
        msisdn = "0912345678",
        isDormant = true,
        hsdConfirmedDate = d("2027-02-01"),
        hsdConfirmedAt = d("2026-08-15"),
        identityVerified = TriState.NO,
        retentionPackage = "VTVANG",
    )

    private val items = listOf(
        TrackedItem(
            id = "hsd", name = "Hạn số", groupId = "sim1", kind = Kind.PREPAID_SIM,
            expiresOn = d("2027-02-01"), actByOffsetDays = 7, anchorDate = d("2027-02-01"),
            dateSource = DateSource.USER_CONFIRMED, lastVerifiedAt = d("2026-08-15"),
        ),
        TrackedItem(
            id = "netflix", name = "Netflix", kind = Kind.RECURRING,
            expiresOn = d("2026-09-01"), anchorDate = d("2026-01-01"), cycle = Cycle.MONTHLY,
            amountMinor = 231_000, currency = "VND",
            actionUrl = "https://netflix.com/cancelplan", actionLabel = "Hủy Netflix",
            note = "Hủy phải vào web, trong app không hủy được",
        ),
        TrackedItem(
            id = "claude", name = "Claude", kind = Kind.RECURRING,
            expiresOn = d("2026-09-05"), anchorDate = d("2026-02-05"), cycle = Cycle.MONTHLY,
            amountMinor = 2000, currency = "USD",
        ),
    )

    private val history = listOf(
        HandledEvent(
            id = "e1", itemId = "claude", handledAtEpochSeconds = 1_755_000_000,
            forDueDate = d("2026-08-05"),
            amountMinor = 2000, currency = "USD",
            fxRateScaled = 260_460_000, fxRateScale = 4,
            fxRateDate = d("2026-08-04"), fxSource = "bundled",
            baseAmountMinor = 520_920, actualChargedMinor = 532_745,
        ),
    )

    private fun sampleFile() = Backup.build("2026-08-15T10:00:00Z", items, listOf(group), listOf(sim), history)

    // A backup that cannot be restored is not a backup. This is the whole point
    // of the file.
    @Test
    fun `a full round trip restores every record`() = runTest {
        val encoded = Backup.encode(sampleFile())
        val decoded = Backup.decode(encoded).getOrThrow()

        val target = repo()
        val report = BackupRestore(target).restore(decoded, nowEpochSeconds = 1_755_000_100)

        assertTrue(report.isClean, "skipped: ${report.skipped}")
        assertEquals(3, report.items)
        assertEquals(1, report.groups)
        assertEquals(1, report.sims)
        assertEquals(1, report.events)

        assertEquals(3, target.observeAll().first().size)
    }

    @Test
    fun `restored items keep every field that matters`() = runTest {
        val target = repo()
        BackupRestore(target).restore(Backup.decode(Backup.encode(sampleFile())).getOrThrow(), 1)

        val netflix = target.observeById("netflix").first()!!
        assertEquals("Netflix", netflix.name)
        assertEquals(Cycle.MONTHLY, netflix.cycle)
        assertEquals(231_000, netflix.amountMinor)
        assertEquals("VND", netflix.currency)
        assertEquals("https://netflix.com/cancelplan", netflix.actionUrl)
        assertEquals("Hủy phải vào web, trong app không hủy được", netflix.note)
        assertEquals(d("2026-01-01"), netflix.anchorDate)
    }

    // Provenance must survive: a date the user confirmed with their carrier and
    // one typed from memory cannot come back looking alike.
    @Test
    fun `date provenance survives the round trip`() = runTest {
        val target = repo()
        BackupRestore(target).restore(Backup.decode(Backup.encode(sampleFile())).getOrThrow(), 1)

        val hsd = target.observeById("hsd").first()!!
        assertEquals(DateSource.USER_CONFIRMED, hsd.dateSource)
        assertEquals(d("2026-08-15"), hsd.lastVerifiedAt)
        assertEquals(7, hsd.actByOffsetDays)
    }

    @Test
    fun `the FX snapshot survives so history stays frozen`() = runTest {
        val target = repo()
        BackupRestore(target).restore(Backup.decode(Backup.encode(sampleFile())).getOrThrow(), 1)

        val event = target.observeHistory("claude").first().single()
        assertEquals(520_920, event.baseAmountMinor)
        assertEquals("bundled", event.fxSource)
        assertEquals(d("2026-08-04"), event.fxRateDate)
        assertEquals(532_745, event.actualChargedMinor)
    }

    @Test
    fun `SIM state including the three clocks survives`() = runTest {
        val target = repo()
        BackupRestore(target).restore(Backup.decode(Backup.encode(sampleFile())).getOrThrow(), 1)

        val restored = target.observeSims().first().single()
        assertEquals(Carrier.VIETTEL, restored.carrier)
        assertTrue(restored.isDormant)
        assertEquals(d("2027-02-01"), restored.hsdConfirmedDate)
        assertEquals(d("2026-08-15"), restored.hsdConfirmedAt)
        assertEquals(TriState.NO, restored.identityVerified)
        assertEquals("VTVANG", restored.retentionPackage)
    }

    @Test
    fun `an unrecognised identity value restores as unknown, never as verified`() = runTest {
        val file = sampleFile().let { it.copy(sims = it.sims.map { s -> s.copy(identityVerified = "MAYBE") }) }
        val target = repo()
        BackupRestore(target).restore(file, 1)

        assertEquals(TriState.UNKNOWN, target.observeSims().first().single().identityVerified)
    }

    @Test
    fun `a corrupt file fails rather than restoring garbage`() {
        assertTrue(Backup.decode("not json at all").isFailure)
        assertTrue(Backup.decode("{}").isFailure)
    }

    @Test
    fun `a file from a newer app version is refused with an explanation`() {
        val future = Backup.encode(sampleFile()).replace("\"schemaVersion\": 1", "\"schemaVersion\": 99")
        val result = Backup.decode(future)

        assertTrue(result.isFailure)
        assertContains(result.exceptionOrNull()?.message.orEmpty(), "phiên bản mới hơn")
    }

    @Test
    fun `unknown fields are ignored so a newer file still restores`() {
        val withExtra = Backup.encode(sampleFile()).replaceFirst("{", """{"futureField": "whatever",""")
        assertTrue(Backup.decode(withExtra).isSuccess)
    }

    @Test
    fun `a partial file restores what it can and reports the rest`() = runTest {
        val broken = sampleFile().let {
            it.copy(items = it.items + it.items.first().copy(id = "bad", expiresOn = "not-a-date"))
        }
        val target = repo()
        val report = BackupRestore(target).restore(broken, 1)

        assertEquals(3, report.items, "the good rows still land")
        assertTrue(report.skipped.isNotEmpty(), "the bad row must be reported, not swallowed")
        assertTrue(!report.isClean)
    }

    // ---- CSV ----

    @Test
    fun `CSV has a header and one row per item`() {
        val lines = Backup.toCsv(items).lines()
        assertEquals(4, lines.size)
        assertTrue(lines.first().startsWith("id,name,group_id"))
    }

    // A spreadsheet would read "231.000" as a decimal and silently divide a
    // Vietnamese amount by a thousand, so amounts stay in minor units.
    @Test
    fun `CSV writes raw minor units with the currency in its own column`() {
        val netflixRow = Backup.toCsv(items).lines().first { it.startsWith("netflix") }
        assertContains(netflixRow, "231000")
        assertContains(netflixRow, "VND")
        assertTrue(!netflixRow.contains("231.000"), "must not be pre-formatted")
    }

    @Test
    fun `CSV escapes commas and quotes`() {
        val tricky = listOf(
            items[1].copy(id = "x", name = """Netflix, "Premium" plan"""),
        )
        val row = Backup.toCsv(tricky).lines().last()
        assertContains(row, """"Netflix, ""Premium"" plan"""")
    }

    @Test
    fun `an empty export is still valid`() {
        val empty = Backup.build("2026-08-15T10:00:00Z", emptyList(), emptyList(), emptyList(), emptyList())
        assertTrue(Backup.decode(Backup.encode(empty)).isSuccess)
        assertEquals(1, Backup.toCsv(emptyList()).lines().size, "header only")
    }
}

File: test/unit/backup/backup_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:subdock/backup/backup.dart';
import 'package:subdock/domain/local_date.dart';
import 'package:subdock/domain/model.dart';
import 'package:subdock/domain/recurrence.dart';

void main() {
  LocalDate d(String iso) => LocalDate.parse(iso);

  final items = [
    TrackedItem(
      id: 'hsd',
      name: 'Hạn số',
      category: Category.subscription,
      expiresOn: d('2027-02-01'),
      actByOffsetDays: 7,
      anchorDate: d('2027-02-01'),
      dateSource: DateSource.userConfirmed,
      lastVerifiedAt: d('2026-08-15'),
    ),
    TrackedItem(
      id: 'netflix',
      name: 'Netflix',
      category: Category.subscription,
      expiresOn: d('2026-09-01'),
      anchorDate: d('2026-01-01'),
      cycle: Cycle.monthly,
      amountMinor: 231000,
      currency: 'VND',
      actionUrl: 'https://netflix.com/cancelplan',
      actionLabel: 'Hủy Netflix',
      note: 'Hủy phải vào web, trong app không hủy được',
    ),
    TrackedItem(
      id: 'claude',
      name: 'Claude',
      category: Category.subscription,
      expiresOn: d('2026-09-05'),
      anchorDate: d('2026-02-05'),
      cycle: Cycle.monthly,
      amountMinor: 2000,
      currency: 'USD',
    ),
  ];

  final history = [
    HandledEvent(
      id: 'e1',
      itemId: 'claude',
      handledAtEpochSeconds: 1755000000,
      forDueDate: d('2026-08-05'),
      amountMinor: 2000,
      currency: 'USD',
      fxRateScaled: 260460000,
      fxRateScale: 4,
      fxRateDate: d('2026-08-04'),
      fxSource: 'bundled',
      baseAmountMinor: 520920,
      actualChargedMinor: 532745,
    ),
  ];

  BackupFile sampleFile() =>
      Backup.build('2026-08-15T10:00:00Z', items, history);

  // A backup that cannot be restored is not a backup. This is the whole point
  // of the file.
  test('a full round trip carries every record', () {
    final decoded = Backup.decode(Backup.encode(sampleFile()));

    expect(decoded.items.length, 3);
    expect(decoded.history.length, 1);
  });

  test('restored items keep every field that matters', () {
    final decoded = Backup.decode(Backup.encode(sampleFile()));
    final netflix = decoded.items.firstWhere((e) => e.id == 'netflix');

    expect(netflix.name, 'Netflix');
    expect(netflix.cycle, 'MONTHLY');
    expect(netflix.amountMinor, 231000);
    expect(netflix.currency, 'VND');
    expect(netflix.actionUrl, 'https://netflix.com/cancelplan');
    expect(netflix.note, 'Hủy phải vào web, trong app không hủy được');
    expect(netflix.anchorDate, '2026-01-01');
  });

  // Provenance must survive: a date the user confirmed with their provider and
  // one typed from memory cannot come back looking alike.
  test('date provenance survives the round trip', () {
    final decoded = Backup.decode(Backup.encode(sampleFile()));
    final hsd = decoded.items.firstWhere((e) => e.id == 'hsd');

    expect(hsd.dateSource, 'USER_CONFIRMED');
    expect(hsd.lastVerifiedAt, '2026-08-15');
    expect(hsd.actByOffsetDays, 7);
  });

  test('the FX snapshot survives so history stays frozen', () {
    final decoded = Backup.decode(Backup.encode(sampleFile()));
    final event = decoded.history.single;

    expect(event.baseAmountMinor, 520920);
    expect(event.fxSource, 'bundled');
    expect(event.fxRateDate, '2026-08-04');
    expect(event.actualChargedMinor, 532745);
  });

  // An older export has `kind` plus a spend `categoryId` and no `category` at
  // all. Restoring it must not quietly reclassify a passport as a
  // subscription, which is what a bare default would do.
  test('a schema-2 file maps its two old fields onto one category', () {
    String category(Map<String, dynamic> json) =>
        BackupItem.categoryFromJson(json);

    expect(
      category({'kind': 'DOCUMENT', 'categoryId': 'documents'}),
      'DOCUMENT',
    );
    expect(
      category({'kind': 'RECURRING', 'categoryId': 'insurance'}),
      'INSURANCE',
    );
    expect(category({'kind': 'RECURRING', 'categoryId': 'home_bills'}), 'BILL');
    expect(
      category({'kind': 'TRIAL', 'categoryId': 'entertainment'}),
      'SUBSCRIPTION',
    );
    expect(
      category({'kind': 'PREPAID', 'categoryId': 'telecom'}),
      'SUBSCRIPTION',
    );
    expect(
      category({'category': 'INSURANCE', 'kind': 'BILL'}),
      'INSURANCE',
      reason: 'a current file is never re-derived',
    );
    expect(() => category({}), throwsFormatException);
  });

  test('a corrupt file fails rather than restoring garbage', () {
    expect(() => Backup.decode('not json at all'), throwsFormatException);
    expect(() => Backup.decode('{}'), throwsFormatException);
  });

  test('a file from a newer app version is refused with an explanation', () {
    final future = Backup.encode(sampleFile())
        .replaceAll('"schemaVersion": 3', '"schemaVersion": 99');

    expect(
      () => Backup.decode(future),
      throwsA(
        isA<FormatException>().having(
          (e) => e.message,
          'message',
          contains('from a newer version'),
        ),
      ),
    );
  });

  test('unknown fields are ignored so a newer file still restores', () {
    final withExtra = Backup.encode(sampleFile())
        .replaceFirst('{', '{"futureField": "whatever",');

    expect(Backup.decode(withExtra).items.length, 3);
  });

  group('CSV', () {
    test('has a header and one row per item', () {
      final lines = Backup.toCsv(items).split('\n');
      expect(lines.length, 4);
      expect(lines.first, startsWith('id,name,category'));
    });

    // A spreadsheet would read "231.000" as a decimal and silently divide a
    // Vietnamese amount by a thousand, so amounts stay in minor units.
    test('writes raw minor units with the currency in its own column', () {
      final row = Backup.toCsv(items)
          .split('\n')
          .firstWhere((l) => l.startsWith('netflix'));

      expect(row, contains('231000'));
      expect(row, contains('VND'));
      expect(
        row,
        isNot(contains('231.000')),
        reason: 'must not be pre-formatted',
      );
    });

    test('escapes commas and quotes', () {
      final tricky = [
        TrackedItem(
          id: 'x',
          name: 'Netflix, "Premium" plan',
          category: Category.subscription,
          expiresOn: d('2026-09-01'),
          anchorDate: d('2026-09-01'),
        ),
      ];
      final row = Backup.toCsv(tricky).split('\n').last;
      expect(row, contains('"Netflix, ""Premium"" plan"'));
    });

    test('an empty export is still valid', () {
      final empty = Backup.build('2026-08-15T10:00:00Z', [], []);
      expect(Backup.decode(Backup.encode(empty)).items, isEmpty);
      expect(Backup.toCsv([]).split('\n').length, 1, reason: 'header only');
    });
  });
}
